File: src/shared/types/ipc.ts
export type AppError = {
  code: string;
  message: string;
  retryable: boolean;
};

export interface IpcResponse<T = unknown> {
  success: boolean;
  data?: T;
  error?: AppError;
}

export function isAppError(value: unknown): value is AppError {
  return (
    typeof value === "object" &&
    value !== null &&
    "code" in value &&
    "message" in value &&
    "retryable" in value
  );
}

File: src-tauri/src/shared/types/ipc.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppError {
    pub code: String,
    pub message: String,
    pub retryable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpcResult<T: Serialize = serde_json::Value> {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<T>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<AppError>,
}

impl<T: Serialize> IpcResult<T> {
    pub fn ok(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
        }
    }

    /// 비즈니스 에러 응답 helper (Ok-Only). 예: features/app 의 app_ping 입력 검증.
    pub fn err(code: &str, message: impl Into<String>, retryable: bool) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(AppError {
                code: code.to_string(),
                message: message.into(),
                retryable,
            }),
        }
    }
}
